packages/nextjs/app/debug/_components/contract/utilsDisplay.tsx

```diff
@@ -24,33 +24,87 @@ export const displayTxResult = (
   if (typeof displayContent === "bigint") {
     try {
       const asNumber = Number(displayContent);
-      if (asNumber <= Number.MAX_SAFE_INTEGER && asNumber >= Number.MIN_SAFE_INTEGER) {
-        return asNumber;
+      if (asText) {
+        if (asNumber <= Number.MAX_SAFE_INTEGER && asNumber >= Number.MIN_SAFE_INTEGER) {
+          return asNumber;
+        } else {
+          return "Ξ" + formatEther(displayContent);
+        }
       } else {
-        return "Ξ" + formatEther(displayContent);
+        return <NumberDisplay value={displayContent} />;
       }
     } catch (e) {
       return "Ξ" + formatEther(displayContent);
     }
   }
 
-  if (typeof displayContent === "string" && isAddress(displayContent)) {
-    return asText ? displayContent : <Address address={displayContent} />;
+  if (typeof displayContent === "string") {
+    if (isAddress(displayContent)) {
+      return asText ? displayContent : <Address address={displayContent} />;
+    }
+
+    if (isHex(displayContent)) {
+      return displayContent; // don't add quotes
+    }
   }
 
   if (Array.isArray(displayContent)) {
-    const mostReadable = (v: DisplayContent) =>
-      ["number", "boolean"].includes(typeof v) ? v : displayTxResultAsText(v);
-    const displayable = JSON.stringify(displayContent.map(mostReadable), replacer);
-
-    return asText ? (
-      displayable
-    ) : (
-      <span style={{ overflowWrap: "break-word", width: "100%" }}>{displayable.replaceAll(",", ",\n")}</span>
-    );
+    return asText ? JSON.stringify(displayContent, replacer) : <TupleDisplay value={displayContent} />;
+  }
+
+  if (typeof displayContent === "object") {
+    return <StructDisplay value={displayContent} />;
   }
 
   return JSON.stringify(displayContent, replacer, 2);
 };
 
-const displayTxResultAsText = (displayContent: DisplayContent) => displayTxResult(displayContent, true);
+const NumberDisplay = ({ value }: { value: bigint }) => {
+  const [isEther, setIsEther] = useState(false);
+
+  const asNumber = Number(value);
+  if (asNumber <= Number.MAX_SAFE_INTEGER && asNumber >= Number.MIN_SAFE_INTEGER) {
```

Comment:
I'm also not sure about this restriction. With this `if` we can't convert let's say `Number.MAX_SAFE_INTEGER` = `9007199254740991` to `0.009...`. But can convert `9007199254740992`.

Reply:
This behavior is in line with previous behavior. I believe this is to ensure that things like lengths or counts, usually small numbers, aren't mistaken for wei.

I think the limit can be changed, however. What would be a good limit? I'm thinking somewhere around 1e9 (one billion) might make sense, but at the same time would be formatted as 0.000000001, which seems like too small a number.

Reply:
> With this if we can't convert let's say Number.MAX_SAFE_INTEGER = 9007199254740991 to 0.009.... But can convert 9007199254740992

Yeah agree, but not sure either whats the best restrictions would be,  I think current restriction looks fine to me for now since previously also it has been this same case

Reply:
ok, let's leave it as is